### DISCUSSION TITLE
Creating interdependent schemas?

### DISCUSSION THREAD

**Discussion by AllySummers** [colinhacks/zod#558] [Q&A]
Lets say for example I have the following schema:
```ts
const schema1 = z.object({
  schema2: schema2.array(),
  schema3: schema3.array()
})

const schema2 = z.object({
  schema1: schema1.array()
  schema3: schema3.array()
})

const schema3 = z.object({
  schema1: schema1.array(),
  schema2: schema2.array()
})
```
Obviously when you define the schemas like that, you will get errors that the variable is not defined.

My current workaround is using functions to return the schema, like this:
```ts
function schema1() {
  return z.object({
    schema2: schema2().array(),
    schema3: schema3().array()
  })
}

function schema2() {
  return z.object({
    schema1: schema1().array()
    schema3: schema3().array()
  })
}

function schema3() {
  return z.object({
    schema1: schema1().array(),
    schema2: schema2().array()
  })
}
```

Is there some other way to implement this that I'm not thinking of?

**Comment by scotttrinh** [comment]
This is similar to [`z.lazy`](https://github.com/colinhacks/zod#recursive-types), but I don't have a ton of experience with using it so I'm not sure if it works outside of the typical (recursive) use case. If you find that it works for your use case, it's probably worth adding some details to our README, so let us know or open a PR and I'll review!

**Comment by colinhacks** [comment]
Clever! This is essentially what `z.lazy` does, but `z.lazy` defers the execution of the functions until it's needed (that is, when you actually try to parse some data with the schema).
